source/drivers/HIDKeyboard.cpp

```diff
@@ -0,0 +1,409 @@
+
+            if(status == DEVICE_OK)
+                keyPressedCountConsumer++;
+            else
+                keyStateConsumer[newIndex] = HID_KEYBOARD_KEY_OFF; //we could not make the change
+
+            break;
+
+        }
+        default:
+            //unknown report
+            status = DEVICE_INVALID_PARAMETER;
+    }
+
+    return status;
+}
+
+int USBHIDKeyboard::keyUp(uint8_t key, uint8_t reportID)
+{
+    int status;
+    int newIndex = -1;
+
+    switch(reportID){
+        case HID_KEYBOARD_REPORT_GENERIC: {
+
+            // ?? hmm no keys are pressed ??
+            if(keyPressedCountGeneric == 0)
+                return DEVICE_INVALID_PARAMETER;
+
+            // try to find the passed key
+            for(int i=HID_KEYBOARD_MODIFIER_OFFSET; i<HID_KEYBOARD_KEYSTATE_SIZE_GENERIC; i++){
+                if(keyStateGeneric[i] == key){
+                    keyStateGeneric[i] = HID_KEYBOARD_KEY_OFF;
+                    newIndex = i;
+                    break;
+                }
+            }
+
+            // the passed key is not pressed
+            if(newIndex == -1)
+                return DEVICE_INVALID_PARAMETER;
+
+            //write to the host
+            uint8_t report[HID_KEYBOARD_KEYSTATE_SIZE_GENERIC + 1] = {reportID};
+            memcpy(report + 1, keyStateGeneric, HID_KEYBOARD_KEYSTATE_SIZE_GENERIC);
+            status = in->write(report, sizeof(report));
+
+            if(status == DEVICE_OK)
+                keyPressedCountGeneric--;
+            else
+                keyStateGeneric[newIndex] = key; //we could not make the change. Revert
+            break;
+        }
+        case HID_KEYBOARD_REPORT_CONSUMER: {
+
+            // ?? hmm no keys are pressed ??
+            if(keyPressedCountConsumer == 0)
+                return DEVICE_INVALID_PARAMETER;
+
+            // try to find the passed key
+            for(int i=HID_KEYBOARD_MODIFIER_OFFSET; i<HID_KEYBOARD_KEYSTATE_SIZE_CONSUMER; i++){
+                if(keyStateConsumer[i] == key){
+                    keyStateConsumer[i] = HID_KEYBOARD_KEY_OFF;
+                    newIndex = i;
+                    break;
+                }
+            }
+
+            // the passed key is not pressed
+            if(newIndex == -1)
+                return DEVICE_INVALID_PARAMETER;
+
+            //write to the host
+            uint8_t report[HID_KEYBOARD_KEYSTATE_SIZE_CONSUMER + 1] = {reportID};
+            memcpy(report + 1, keyStateConsumer, HID_KEYBOARD_KEYSTATE_SIZE_CONSUMER);
+            status = in->write(report, sizeof(report));
+
+            if(status == DEVICE_OK)
+                keyPressedCountConsumer--;
+            else
+                keyStateConsumer[newIndex] = key; //we could not make the change. Revert
+            break;
+        }
+        default:
+            //unknown report
+            status = DEVICE_INVALID_PARAMETER;
+
+    }
+
+    return status;
+}
+
+int USBHIDKeyboard::modifierKeyDown(uint8_t key, uint8_t reportID)
```

Comment:
Is it possible to send a capital 'A' without using the modifier API?

For instance:

```cpp
USBHIDKeyboard keyboard;

keyboard.keyDown('A');
keyboard.keyUp('A');
```

Would that result in a capital A on my screen? Or would it be something more like:
```cpp
USBHIDKeyboard keyboard;

keyboard.modifierKeyDown(LEFT_SHIFT);
keyboard.keyDown('a');
keyboard.keyUp('a');
keyboard.modifierKeyUp(LEFT_SHIFT);
```

Reply:
Also, how does a user set the left shift modifier?

Reply:
the general idea was that you would access the API the same way you would type on a keyboard. the second example would be correct when replaced by the correct keycodes for a standard US keyboard.

```
USBHIDKeyboard keyboard;

keyboard.modifierKeyDown(0x02);
keyboard.keyDown(0x04);
keyboard.keyUp(0x04);
keyboard.modifierKeyUp(0x02);
```